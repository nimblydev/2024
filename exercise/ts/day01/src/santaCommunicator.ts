import { Logger } from "./logger";

interface ReindeerTravelDetails {
  reindeer: Reindeer;
  numberOfDaysBeforeChristmas: number;
}

export class ReindeerLocation {
  private readonly name: string;
  private readonly numberOfDaysFromNorthPole: number;

  constructor(location: string, numberOfDaysFromNorthPole: number) {
    this.name = location;
    this.numberOfDaysFromNorthPole = numberOfDaysFromNorthPole;
  }

  public get getLocation(): string {
    return this.name;
  }
  public get getNumberOfDaysForComingBack(): number {
    return this.numberOfDaysFromNorthPole;
  }
}

export class Reindeer {
  private readonly name: string;
  private location: ReindeerLocation;

  constructor(name: string, currentLocation: ReindeerLocation) {
    this.name = name;
    this.location = currentLocation;
  }
  public get reindeerName(): string {
    return this.name;
  }

  public get currentLocation(): string {
    return this.location.getLocation;
  }

  public set newLocation(location: ReindeerLocation) {
    this.location = location;
  }

  public getNumberOfDaysForComingBack(): number {
    return this.location.getNumberOfDaysForComingBack;
  }
}

export class SantaCommunicator {
  private readonly numberOfDaysToRest: number;

  constructor(numberOfDaysToRest: number) {
    this.numberOfDaysToRest = numberOfDaysToRest;
  }

  public composeMessage({
    reindeer,
    numberOfDaysBeforeChristmas,
  }: ReindeerTravelDetails): string {
    const daysBeforeReturn = this.daysBeforeReturn(
      reindeer.getNumberOfDaysForComingBack(),
      numberOfDaysBeforeChristmas
    );
    return `Dear ${reindeer.reindeerName}, please return from ${reindeer.currentLocation} in ${daysBeforeReturn} day(s) to be ready and rest before Christmas.`;
  }

  public isReindeerOverdue(
    reindeer: Reindeer,
    numberOfDaysBeforeChristmas: number,
    logger: Logger
  ): boolean {
    if (
      this.daysBeforeReturn(
        reindeer.getNumberOfDaysForComingBack(),
        numberOfDaysBeforeChristmas
      ) <= 0
    ) {
      logger.log(
        `Overdue for ${reindeer.reindeerName} located ${reindeer.currentLocation}.`
      );
      return true;
    }
    return false;
  }

  private daysBeforeReturn(
    numberOfDaysForComingBack: number,
    numberOfDaysBeforeChristmas: number
  ): number {
    return (
      numberOfDaysBeforeChristmas -
      numberOfDaysForComingBack -
      this.numberOfDaysToRest
    );
  }
}
