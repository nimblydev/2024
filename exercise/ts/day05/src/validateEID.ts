const EID_LENGTH = 8;
const ELVEN_SEX = ["1", "2", "3"];
const numericStringPattern = /^\d+$/;
const eidModulus = 97;
const CONTROL_KEY_LENGTH = 2;
const ZERO_PADDING_CHAR = "0";

export const validateEID = (eid: string) => {
  const eidYearCode = eid?.substring(1, 3);
  const eidSexIdentifier = eid?.substring(0, 1);
  const eidCheckCharacter = eid?.substring(6, 8);
  const eidPrefix = eid?.substring(0, 6);

  return (
    isNotEmpty(eid) &&
    isLengthValid(eid) &&
    isOnlyDigits(eid) &&
    isValidSexDigit(eidSexIdentifier) &&
    isValidEIDYear(eidYearCode) &&
    isControlKeyValid(eidPrefix, eidCheckCharacter)
  );
};

const isValidSexDigit = (sexDigit: string) => ELVEN_SEX.includes(sexDigit);

const isNotEmpty = (eid: string) => eid !== null && eid !== "";

const isLengthValid = (eid: string) => eid.length === EID_LENGTH;

const isOnlyDigits = (eid: string) => numericStringPattern.test(eid);

const isValidEIDYear = (year: string) =>
  isOnlyDigits(year) && year.length === 2;

const isControlKeyValid = (eidPrefix: string, controlKey: string) => {
  const modulo = parseInt(eidPrefix, 10) % eidModulus;
  const complementedValue = (eidModulus - modulo)
    .toString()
    .padStart(CONTROL_KEY_LENGTH, ZERO_PADDING_CHAR);

  return complementedValue === controlKey;
};
