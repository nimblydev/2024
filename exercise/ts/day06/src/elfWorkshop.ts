export class ElfWorkshop {
  // readability: the class name should be more explicit, like ElfTaskManager
  // best-practice: for the sake of immutability, the taskList should be a private property and should be accessed through a getter
  // best-practice: the taskList should be readonly and initialized in the constructor
  // best-practice: use underscore to indicate that the property is private
  taskList: string[] = [];

  // best practice: where is the constructor ? it should be used to initialize the taskList

  // we could avoid a primitive type for task and use a Task class or a structured object
  addTask(task: string): void {
    if (task !== "") {
      this.taskList.push(task);
    }
  }
  // typing:  completeTask() returns a string but it can also return null
  completeTask(): string {
    // question: no sorting, so the first task added is the first to be removed ?
    if (this.taskList.length > 0) {
      // with an other approach, we could return the first uncompleted task, without testing if the taskList is empty
      // functional: shift() removes the first element, a better approach with a non-primitive Task would be to filter uncompleted tasks and return the first one
      // then change its status to completed
      return this.taskList.shift();
    }
    return null; // functional/question:  why not return an empty string or an explicit message ?
  }
}
