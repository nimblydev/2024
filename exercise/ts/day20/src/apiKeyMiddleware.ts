import { NextFunction, Request, Response } from "express";

export const apiKeyMiddleware =
  (validApiKey: string) =>
  (req: Request, res: Response, next: NextFunction) => {
    const apiKey = req.headers["x-api-key"];

    if (apiKey && apiKey === validApiKey) {
      next();
    } else {
      res.status(403).send("Forbidden: Invalid API Key");
    }
  };
