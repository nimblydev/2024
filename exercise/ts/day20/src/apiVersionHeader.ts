import { NextFunction, Request, Response } from "express";

const apiVersionHeaderKey = "x-api-version";
export const apiVersionHeaderMiddleware =
  (version: string) => (req: Request, res: Response, next: NextFunction) => {
    res.setHeader(apiVersionHeaderKey, version);
    next();
  };
