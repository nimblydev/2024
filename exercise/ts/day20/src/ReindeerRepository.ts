import { Either } from "fp-ts/lib/Either";
import { Reindeer, ReindeerErrorCode } from "./types";
import { ReindeerToCreate } from "./service";

export interface ReindeerRepository {
  get(id: string): Either<ReindeerErrorCode, Reindeer>;
  create(
    reindeerToCreate: ReindeerToCreate
  ): Either<ReindeerErrorCode, Reindeer>;
}
