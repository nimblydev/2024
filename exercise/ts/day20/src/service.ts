import { Either, right } from "fp-ts/lib/Either";
import { Reindeer, ReindeerColor, ReindeerErrorCode } from "./types";
import { ReindeerRepository } from "./ReindeerRepository";

export class ReindeerService {
  constructor(private readonly reindeerRepository: ReindeerRepository) {}

  public health(): Either<never, string> {
    return right("OK");
  }

  public get(id: string): Either<ReindeerErrorCode, Reindeer> {
    return this.reindeerRepository.get(id);
  }

  public create(
    reindeerToCreate: ReindeerToCreate
  ): Either<ReindeerErrorCode, Reindeer> {
    return this.reindeerRepository.create(reindeerToCreate);
  }
}

export class ReindeerToCreate {
  constructor(public name: string, public color: ReindeerColor) {}
}
