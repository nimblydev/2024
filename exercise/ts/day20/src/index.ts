import express from "express";
import { ReindeerService, ReindeerToCreate } from "./service";
import { fold } from "fp-ts/Either";
import { ReindeerErrorCode } from "./types";
import { apiKeyMiddleware } from "./apiKeyMiddleware";
import { InMemoryReinderRepository } from "../tests/fakes/InMemoryReinderRepository";
import { apiVersionHeaderMiddleware } from "./apiVersionHeader";

export const API_KEY = "123e4567-e89b-12d3-a456-426614174000";
const API_VERSION = "1.0";

export const app = express();
app.use(express.json());

const reindeerService = new ReindeerService(new InMemoryReinderRepository());

app.use(apiVersionHeaderMiddleware(API_VERSION));
app.use(apiKeyMiddleware(API_KEY));

app.get("/reindeer/:id", (req, res) => {
  const result = reindeerService.get(req.params.id);
  fold(
    (error) => {
      if (error === ReindeerErrorCode.NotFound) {
        res.status(404).send("Reindeer not found");
      }
    },
    (reindeer) => res.status(200).json(reindeer)
  )(result);
});

app.post("/reindeer", (req, res) => {
  const { name, color } = req.body;
  const reindeerToCreate: ReindeerToCreate = { name, color };
  const result = reindeerService.create(reindeerToCreate);
  fold(
    (error) => {
      if (error === ReindeerErrorCode.AlreadyExist) {
        res.status(409).send("Reindeer already exists");
      }
    },
    (reindeer) => res.status(201).json(reindeer)
  )(result);
});

const port = 3000;
app.listen(port, () => {
  console.log(`Server is running on http://localhost:${port}`);
});
