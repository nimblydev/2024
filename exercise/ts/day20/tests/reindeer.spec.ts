import { Server } from "node:http";
import { v4 as uuidv4 } from "uuid";
import { app } from "../src";
import { ReindeerColor } from "../src/types";
import { AddressInfo } from "net";

describe("Reindeer API", () => {
  let server: Server;
  let baseUrl: string;
  beforeAll((done) => {
    server = app.listen(() => {
      const { port } = server.address() as AddressInfo;
      baseUrl = `http://localhost:${port}`;
      console.log(`Server started at ${baseUrl}`);
      done();
    });
  });

  afterAll((done) => {
    server.closeAllConnections();
    server.close(() => {
      console.log("Server stopped");
      done();
    });
  });
  describe("when using a valid API key", () => {
    const VALID_API_KEY = "123e4567-e89b-12d3-a456-426614174000";
    it("should return an x-api-version header", async () => {
      const response = await fetch(
        `${baseUrl}/reindeer/40f9d24d-d3e0-4596-adc5-b4936ff84b19`
      );
      const apiVersion = response.headers.get("x-api-version");

      expect(apiVersion).toMatchSnapshot("api-version");
    });

    it("should get a reindeer", async () => {
      const response = await fetch(
        `${baseUrl}/reindeer/40f9d24d-d3e0-4596-adc5-b4936ff84b19`,
        {
          headers: { "X-API-Key": VALID_API_KEY },
        }
      );
      const responseBody = await response.json();
      expect(responseBody).toMatchSnapshot("get-reindeer");
      expect(response.status).toBe(200);
    });

    it("should return not found for non-existing reindeer", async () => {
      const nonExistingReindeer = uuidv4();
      const response = await fetch(
        `${baseUrl}/reindeer/${nonExistingReindeer}`,
        {
          headers: { "X-API-Key": VALID_API_KEY },
        }
      );

      expect(response.status).toBe(404);
    });

    it("should create a reindeer", async () => {
      const requestPayload = {
        name: "Rudolph",
        color: ReindeerColor.Purple,
      };
      const response = await fetch(`${baseUrl}/reindeer`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "x-api-key": VALID_API_KEY,
        },
        body: JSON.stringify(requestPayload),
      });
      const responseBody = await response.json();

      expect(responseBody).toMatchSnapshot("create-reindeer");
      expect(response.status).toBe(201);
    });

    it("should return conflict when trying to create an existing reindeer", async () => {
      const requestPayload = {
        name: "Petar",
        color: ReindeerColor.Purple,
      };
      const response = await fetch(`${baseUrl}/reindeer`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-API-Key": VALID_API_KEY,
        },
        body: JSON.stringify(requestPayload),
      });

      expect(response.status).toBe(409);
    });
  });

  describe("when useing a invalid API key", () => {
    it("should return forbidden when trying to get a reindeer", async () => {
      const response = await fetch(
        `${baseUrl}/reindeer/40f9d24d-d3e0-4596-adc5-b4936ff84b19`,
        {
          headers: { "x-api-key": "invalid-api-key" },
        }
      );

      expect(response.status).toBe(403);
    });

    it("should return forbidden when trying to create a reindeer", async () => {
      const requestPayload = {
        name: "Rudolph",
        color: ReindeerColor.Purple,
      };
      const response = await fetch(`${baseUrl}/reindeer`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "x-api-key": "invalid-api-key",
        },
        body: JSON.stringify(requestPayload),
      });

      expect(response.status).toBe(403);
    });
  });
});
