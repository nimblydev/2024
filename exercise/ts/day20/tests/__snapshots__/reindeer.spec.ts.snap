// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reindeer API when using a valid API key should create a reindeer: create-reindeer 1`] = `
{
  "color": 2,
  "id": "19958282-a3b9-5726-b232-db61d5395e22",
  "name": "Rudolph",
}
`;

exports[`Reindeer API when using a valid API key should get a reindeer: get-reindeer 1`] = `
{
  "color": 2,
  "id": "40f9d24d-d3e0-4596-adc5-b4936ff84b19",
  "name": "Petar",
}
`;

exports[`Reindeer API when using a valid API key should return an x-api-version header: api-version 1`] = `"1.0"`;
